microsoft/MixedRealityToolkit-Unity
# Add SelectFar, Clicked and Toggle states to Interactive Element

## Overview

Added support for the following states:

#### SelectFar  
- Selection of an object via far interaction 

|SelectFar State Behavior| SelectFar State Global option behavior |
|---|---|
| ![SelectFarState](https://user-images.githubusercontent.com/53493796/101115405-2b23ce00-3598-11eb-8784-5cf643885e62.gif)| ![SelectFarGlobalProperty](https://user-images.githubusercontent.com/53493796/101222366-7abcd500-363e-11eb-9624-bb35c4a0c686.gif)|


#### ToggleOn and ToggleOff 
- The ToggleOn and ToggleOff states together create a toggle button.  These states are set by a far interaction click by default but can be set from any entry point using the `SetToggleStates()` method 

![ToggleStates](https://user-images.githubusercontent.com/53493796/101115615-7ccc5880-3598-11eb-9147-f106bea13db6.gif)


#### Clicked 
- This state is trigged with a far interaction click by default but can be triggered from any entry point using the `TriggerClickedState()` method.

The following figure illustrates the expected entry points for the clicked state moving forward.  The grey states in the figure are not in this PR. 

![image](https://user-images.githubusercontent.com/53493796/101196471-9e1f5a00-3615-11eb-8939-6ca99d76ddf5.png)


#### Other Updates

- Added tests for the new states
- Added new active state highlight in the inspector


## Changes
- Fixes: #8883 

## To do

- [x] Add clicked state diagram 


## Verification
1. Checkout the branch
1. Open new unity scene
1. Add MRTK to the scene
1. Add a cube to the scene
1. Attach the Interactive Element component
1. Select Add Core State button in the Interactive Element inspector
1. Add the SelectFar state
1. Press play and observe the state value changes in the inspector
